OPTIMIZE DEDUPLICATE BY COLUMNS

Enmk opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Extended `OPTIMIZE ... DEDUPLICATE` syntax to allow explicit (or implicit with asterisk/column transformers) list of columns to check for duplicates on.
...


Detailed description / Documentation draft:
Following syntax variants are now supported:
``` SQL
OPTIMIZE TABLE table DEDUPLICATE; -- the old one
OPTIMIZE TABLE table DEDUPLICATE BY *; -- not the same as the old one, excludes MATERIALIZED columns (see the note below)
OPTIMIZE TABLE table DEDUPLICATE BY * EXCEPT colX;
OPTIMIZE TABLE table DEDUPLICATE BY * EXCEPT (colX, colY);
OPTIMIZE TABLE table DEDUPLICATE BY col1,col2,col3;
OPTIMIZE TABLE table DEDUPLICATE BY COLUMNS('column-matched-by-regex');
OPTIMIZE TABLE table DEDUPLICATE BY COLUMNS('column-matched-by-regex') EXCEPT colX;
OPTIMIZE TABLE table DEDUPLICATE BY COLUMNS('column-matched-by-regex') EXCEPT (colX, colY);
```
**Note** that `*` behaves just like in `SELECT`: `MATERIALIZED`, and `ALIAS` columns are not used for expansion.
Also, it is an error to specify empty list of columns, or write an expression that results in an empty list of columns, or deduplicate by an `ALIAS` column.
Column transformers other than `EXCEPT` are not supported.

Please see tests for examples.
...

Enmk committed:
OPTIMIZE DEDUPLICATE BY columns

Extended OPTIMIZE ... DEDUPLICATE syntax to allow explicit (or implicit with asterisk/column transformers) list of columns to check for duplicates on.

Following syntax variants are now supported:

OPTIMIZE TABLE table DEDUPLICATE; -- the old one
OPTIMIZE TABLE table DEDUPLICATE BY *;
OPTIMIZE TABLE table DEDUPLICATE BY * EXCEPT colX;
OPTIMIZE TABLE table DEDUPLICATE BY * EXCEPT (colX, colY);
OPTIMIZE TABLE table DEDUPLICATE BY col1,col2,col3;
OPTIMIZE TABLE table DEDUPLICATE BY COLUMNS('column-matched-by-regex');
OPTIMIZE TABLE table DEDUPLICATE BY COLUMNS('column-matched-by-regex') EXCEPT colX;
OPTIMIZE TABLE table DEDUPLICATE BY COLUMNS('column-matched-by-regex') EXCEPT (colX, colY);

Note that * behaves just like in SELECT: MATERIALIZED, and ALIAS columns are not used for expansion.
Also, it is an error to specify empty list of columns, or write an expression that results in an empty list of columns, or deduplicate by an ALIAS column.
Column transformers other than EXCEPT are not supported.

Enmk committed:
Fixed parsing invalid cases: prohibit empty lists and APPLY/REPLACE column transformers

Enmk committed:
Fixed and refined unite-test

* no more undefined values for attributes in ReplicatedMergeTreeLogEntry
* validation of string serialization format

Enmk committed:
Updated tests

Enmk committed:
Minor: cleanup

filimonov commented on tests/queries/0_stateless/01581_deduplicate_by_columns_replicated.sql:
```diff
@@ -0,0 +1,55 @@
+DROP TABLE replicated_deduplicate_by_columns_r2;
+
+SYSTEM DROP REPLICA '/clickhouse/tables/test_01581/replicated_deduplicate/replicas/r1';
+SYSTEM DROP REPLICA '/clickhouse/tables/test_01581/replicated_deduplicate/replicas/r2';
```
Normally it's not needed (drop table cleans the zookeeper state), I guess you were hitting some issues only when you kill the test in the middle.

azat replied:
>drop table cleans the zookeeper state

Not for testkeeper `zookeeper.implementation`.
But `SYSTEM DROP REPLICA` also will do nothing in this case.

Enmk replied:
Ok, removed it

Enmk committed:
Fixed building tests with GCC-10

Also a minor cleanup of the test code.

filimonov commented on src/Interpreters/InterpreterOptimizeQuery.cpp:
```diff
@@ -27,7 +32,29 @@ BlockIO InterpreterOptimizeQuery::execute()
+        }
+
+        metadata_snapshot->check(column_names, NamesAndTypesList{}, table_id);
+        for (const auto & primary_key : metadata_snapshot->getPrimaryKeyColumns())
```
https://github.com/filimonov/ClickHouse/blob/7e3524caa1cbdf65f686eaaf6d39049734f5ed4a/src%2FStorages%2FMergeTree%2FMergeTreeData.cpp#L1418-L1452

filimonov replied:
The safest option would be to forbid also partition key and all sorting key column (primary key may have less columns than sorting key).

Enmk replied:
Checking that all keys from `ORDER BY` or `PRIMARY KEY` are present in expanded `DEDUPLICATE BY` columns list, added comment and proper exception message there, updated tests to ensure this case.

Enmk committed:
Fixed test to be less flaky

Also logging expanded list of columns passed from `DEDUPLICATE BY` to actual deduplication routines.

Enmk committed:
Enforcing all sorting keys to be present in DEDUPLICATE BY columns

Updated test and minor cleanup

filimonov commented on src/Interpreters/InterpreterOptimizeQuery.cpp:
```diff
@@ -27,7 +32,38 @@ BlockIO InterpreterOptimizeQuery::execute()
+                        " Expanded deduplicate columns expression: ['{}']",
+                        sorting_key, fmt::join(column_names, "', '"));
+        }
+    }
```
And what's with partition key columns?

Enmk replied:
Partition key is part of sorting key, there is also a case for the in the tests:

https://github.com/ClickHouse/ClickHouse/pull/17846/files#diff-c91002f2000111398950b55920142b1aebc6711be2f3fc195b84f79e57568f80R30

Enmk replied:
Good catch, fixed

Enmk committed:
Checking that columns from PARTITION BY are present in DEDUPLICATE BY

Enmk committed:
Fixed compilation

alexey-milovidov commented on src/Storages/MergeTree/ReplicatedMergeTreeLogEntry.cpp:
```diff
@@ -184,6 +212,21 @@ void ReplicatedMergeTreeLogEntryData::readText(ReadBuffer & in)
+                    for (;;)
+                    {
+                        String tmp_column_name;
+                        readJSONString(tmp_column_name, in);
```
This is quite dangerous, because JSON does not support non-unicode data while we support arbitrary bytes for column names.

alexey-milovidov replied:
This will lead to idiosynchrasy as two different escaping methods will be used in the same file.
Let's reuse existing "tsv-escaped" method.

Enmk replied:
Well, I've done some research, and it looks like the method we use with `TabSeparatedWithNames` (I believe this is what you meant by 'tsv-secaped') wouldn't cut it:
INPUT as stated in C++ code:
```
{"name with space", "\"column\"", "'column'", "колонка", "\u30ab\u30e9\u30e0", "\x00\x01\x03 column \x10\x11\x12"},
```

writeEscapedString
```
deduplicate_by_columns: [name with space,"column",\'column\',колонка,カラム,\0 column ]
```

writeJSONString
```
deduplicate_by_columns: ["name with space","\"column\"","'column'","колонка","カラム","\u0000\u0001\u0003 column \u0010\u0011\u0012"]
```

Please note how `writeEscapedString` eats all non-unicode binary.

alexey-milovidov replied:
> Please note how writeEscapedString and writeCSV eats all non-unicode binary.

No, they don't eat binary data. The bytes are written but they are lost in copy-paste.
Actually our JSON format is also binary safe, but it can be misleading to applications.

alexey-milovidov replied:
```
$ clickhouse-client --query "SELECT '\x00\x01\x02\x03'"
\0
$ clickhouse-client --query "SELECT '\x00\x01\x02\x03'" | xxd
00000000: 5c30 0102 030a                           \0....
```

Enmk replied:
Agree, my bad. Changed that to CSV format, since that makes it easier to parse list of columns. Right now serialized list of columns from the unit tests like this:
```
...
deduplicate_by_columns: "name with space","""column""","'column'","колонка","カラム"," column "
...
```

However, I had to ditch using `\0` in test data since it makes impossible to validate result with `regex`.

filimonov commented:
| ```
Test name | Test status | Test time, sec.
-- | -- | 
test_materialize_mysql_database/test.py::test_select_without_columns_5_7[clickhouse_node1] | FAIL | 189.75
test_materialize_mysql_database/test.py::test_select_without_columns_8_0[clickhouse_node1
```

Broken in master

Enmk committed:
Using CSV-like strings for list of columns to deduplicate by instead of JSON-like notation.

Enmk committed:
Single quotes around column names

alexey-milovidov committed:
Merge branch 'master' into Enmk-Optimize_deduplicate
